core/src/wsv.rs

```diff
@@ -279,23 +279,155 @@ impl WorldStateView {
     /// you likely have data corruption.
     /// - If trigger execution fails
     /// - If timestamp conversion to `u64` fails
-    #[iroha_futures::telemetry_future]
-    #[log(skip(self, block))]
     #[allow(clippy::expect_used)]
-    pub async fn apply(&self, block: VersionedCommittedBlock) -> Result<()> {
+    pub fn apply(&mut self, block: VersionedCommittedBlock) -> Result<()> {
         let time_event = self.create_time_event(block.as_v1())?;
         self.produce_event(Event::Time(time_event));
 
-        self.execute_transactions(block.as_v1()).await?;
+        self.execute_transactions(block.as_v1())?;
 
         self.world.triggers.handle_time_event(time_event);
 
+        /*
+        was previously the following:
         let res = self
             .world
             .triggers
             .inspect_matched(|action, event| -> Result<()> { self.process_trigger(action, event) })
             .await;
 
+        now is: */
+        let res = {
+            let (succeed, res) = {
+                let mut succeed = Vec::new();
+                let mut errors = Vec::new();
+
+                // Cloning and clearing `self.ids_write` so that `handle_` call won't deadlock
+                let mut matched_ids = Vec::new();
+                std::mem::swap(&mut matched_ids, &mut self.world.triggers.matched_ids);
+
+                for (event_type, id) in matched_ids {
+                    // Ignoring `None` variant cause this means that action was deleted after `handle_`
+                    // call and before `inspect_matching()` call
+                    let _ = match event_type {
+                        Event::Data(_) => {
+                            let dangerous_reference = unsafe {
+                                let ptr = &self.world.triggers as *const TriggerSet; // TODO: REALLY FIX THIS
+                                &*ptr
+                            };
+
+                            dangerous_reference.data_triggers.get(&id).map(|entry| {
+                                let action = entry.value();
+                                if let Repeats::Exactly(atomic) = action.repeats() {
+                                    if atomic.get() == 0 {
+                                        return;
+                                    }
+                                }
+
+                                match self.process_trigger(action, event_type) {
+                                    Ok(()) => succeed.push(id),
+                                    Err(err) => errors.push(err),
+                                }
+                            })
+                        }
+                        Event::Pipeline(_) => {
+                            let dangerous_reference = unsafe {
+                                let ptr = &self.world.triggers as *const TriggerSet; // TODO: REALLY FIX THIS
```

Comment:
OK. This comment tells us that something is wrong and the fact that it' bad can be undestood from `unsafe`. Imagine that you're not the one fixing it. An issue number would be just as effective and way more informative.

Reply:
I will create the issue.

Reply:
https://github.com/hyperledger/iroha/issues/2757

Reply:
I think we should change approach before merging with dev branch.

Using `mem::{take, replace}` could be way to go. 

```rust
// Take triggers
let triggers = core::mem::take(&mut self.world.triggers);
for (event_type, id) in matched_ids {
    // ...
}
// Return triggers onto its place
core::mem::replace(&mut self.world.triggers, triggers);
```

Reply:
you are correct, this is definitely UB. `mem::take` or `mem::replace` should be used

Reply:
what you end up with is something like the following
```rs
WorldStateView::process_trigger(WorldStateView *restrict self, *dyn ActionTrait action, Event event_type);
```

but you break the contract of the `*restrict` pointer because your two pointers alias (`action` is a pointer into `self`). This is the definition of UB

Reply:
I would highly recommend this is fixed before merging

Reply:
Small showcase of what happens ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a7d973f922e6e3e7d06e73edd01e7f12)), you could run `miri` to see that it complains on UB.

Reply:
UB is unacceptable.

Reply:
> WTH is going on in this function?? Why entire implementation from `trigger::Set` is here now?

I found it difficult to understand what was going on when it was split up in many small functions. It can probably be broken up again though.

Reply:
As for the UB. I would like to say that the actual operations performed on the world state view are the exact same using the exact same data structures. As far as I know the code doesn't do anything different just because you said one of the references was mut. It enables you to do UB things, but we don't currently. The code is the exact same as if it were 2 immutable references.

Reply:
Problem with this code is that it violates some compiler invariants and because of this it is UB.